src/core/mutatus.scala

```diff
@@ -294,21 +294,27 @@ object Decoder extends Decoder_1 {
     *  This is a suboptimal implementation, and a better solution may be possible with more work
    **/
   def dispatch[T](st: SealedTrait[Decoder, T]): Decoder[T] =
-    value => {
-      value match {
-        case entityValue: EntityValue if entityValue.get().contains("type") =>
-          val entity = entityValue.get()
-          val typeName = entity.getString("type")
-          st.subtypes.collectFirst {
-            case subtype if subtype.typeName.full == typeName =>
-              subtype.typeclass.decodeValue(entity.getValue("value"))
+    encodedValue => {
+      // Naive approach used when fetching entity not indexed by mutatus
+      def firstSucessfull(value: Value[_]) = {
+        st.subtypes.toStream
+          .map { subtype =>
+            Try(subtype.typeclass.decodeValue(value))
           }
-        case value => // Naive approach used when fetching entity not indexed by mutatus
-          st.subtypes.toStream
-            .map { subtype =>
-              Try(subtype.typeclass.decodeValue(value))
+          .collectFirst { case Success(value) => value }
+      }
+
+      encodedValue match {
+        case entityValue: EntityValue =>
+          Try(entityValue.get.getEntity("_meta").getString("typename"))
+            .map { typeName =>
+              st.subtypes.collectFirst {
+                case subtype if subtype.typeName.full == typeName =>
```

Comment:
Do we need to use the full typename, or would it be sufficient just to use the class name? I know that it would be possible to have two types in the same file with the same name, inside different objects, both of which inherit from the same sealed trait... but I also think it would be reasonable not to support that case. The benefit would be that the contents of the database would be closer to a "natural" encoding.

One other possibility would be to use the "full" `typeName` only in the case where there would be an ambiguity, which you could determine by checking the set of `subtypes` in the `SealedTrait` type. While that wouldn't be resilient to changes to the data definitions if they were to rename classes... in general these changes will always break things, so it's not worth trying to accommodate them anyway.

Reply:
I think class name on it's own may be better in such case. If package of sealed trait would change it would need as to decode it using naive method of checking all possible case. In case if all implementations of sealed trait are case objects (eq. enums) it would take first successfully decoded... which would be the same for each entity. It may create huge mess in app.